script/src/syscalls/load_header.rs

```diff
@@ -2,48 +2,75 @@ use crate::syscalls::{
     utils::store_data, Source, SourceEntry, INDEX_OUT_OF_BOUND, ITEM_MISSING,
     LOAD_HEADER_SYSCALL_NUMBER, SUCCESS,
 };
-use ckb_core::cell::ResolvedOutPoint;
+use crate::DataLoader;
+use ckb_core::cell::CellMeta;
 use ckb_core::header::Header;
 use ckb_protocol::Header as FbsHeader;
 use ckb_vm::{
     registers::{A0, A3, A4, A7},
     Error as VMError, Register, SupportMachine, Syscalls,
 };
 use flatbuffers::FlatBufferBuilder;
+use numext_fixed_hash::H256;
 
 #[derive(Debug)]
-pub struct LoadHeader<'a> {
-    resolved_inputs: &'a [ResolvedOutPoint],
-    resolved_deps: &'a [ResolvedOutPoint],
+pub struct LoadHeader<'a, DL> {
+    data_loader: &'a DL,
+    // This can only be used for liner search
+    header_deps: &'a [H256],
+    resolved_inputs: &'a [CellMeta],
+    resolved_cell_deps: &'a [CellMeta],
     group_inputs: &'a [usize],
 }
 
-impl<'a> LoadHeader<'a> {
+impl<'a, DL: DataLoader + 'a> LoadHeader<'a, DL> {
     pub fn new(
-        resolved_inputs: &'a [ResolvedOutPoint],
-        resolved_deps: &'a [ResolvedOutPoint],
+        data_loader: &'a DL,
+        header_deps: &'a [H256],
+        resolved_inputs: &'a [CellMeta],
+        resolved_cell_deps: &'a [CellMeta],
         group_inputs: &'a [usize],
-    ) -> LoadHeader<'a> {
+    ) -> LoadHeader<'a, DL> {
         LoadHeader {
+            data_loader,
+            header_deps,
             resolved_inputs,
-            resolved_deps,
+            resolved_cell_deps,
             group_inputs,
         }
     }
 
-    fn fetch_header(&self, source: Source, index: usize) -> Result<&Header, u8> {
+    fn load_header(&self, cell_meta: &CellMeta) -> Option<Header> {
+        let block_hash = &cell_meta
+            .block_info
+            .as_ref()
+            .expect("block_info of CellMeta should exists when load_header in syscall")
+            .hash;
+        if self.header_deps.iter().any(|hash| hash == block_hash) {
+            self.data_loader.get_header(block_hash)
+        } else {
+            None
+        }
+    }
+
+    fn fetch_header(&self, source: Source, index: usize) -> Result<Header, u8> {
         match source {
             Source::Transaction(SourceEntry::Input) => self
                 .resolved_inputs
                 .get(index)
                 .ok_or(INDEX_OUT_OF_BOUND)
-                .and_then(|r| r.header().ok_or(ITEM_MISSING)),
+                .and_then(|cell_meta| self.load_header(cell_meta).ok_or(ITEM_MISSING)),
             Source::Transaction(SourceEntry::Output) => Err(INDEX_OUT_OF_BOUND),
-            Source::Transaction(SourceEntry::Dep) => self
-                .resolved_deps
+            Source::Transaction(SourceEntry::CellDep) => self
```

Comment:
return INDEX_OUT_OF_BOUND directly.

Reply:
No, it's on purpose, otherwise `dao` contract will not work.

Reply:
It means load the header where the out point located.